controllers/yawol-controller/loadbalancerset/loadbalancerset_controller.go

```diff
@@ -175,39 +175,49 @@ func (r *LoadBalancerSetReconciler) reconcileStatus(
 func (r *LoadBalancerSetReconciler) reconcileReplicas(
 	ctx context.Context,
 	set *yawolv1beta1.LoadBalancerSet,
-	deletedMachines, notReadyMachines, readyMachines []yawolv1beta1.LoadBalancerMachine,
+	machines []yawolv1beta1.LoadBalancerMachine,
+	deletedMachineCount int,
 ) (ctrl.Result, error) {
-	var currentReplicas, desiredReplicas, deletedReplicas int
-	desiredReplicas = set.Spec.Replicas
-	currentReplicas = len(deletedMachines) + len(notReadyMachines) + len(readyMachines)
-	deletedReplicas = len(deletedMachines)
-
-	if currentReplicas < desiredReplicas {
+	if len(machines) < set.Spec.Replicas {
 		if err := r.createMachine(ctx, set); err != nil {
 			return ctrl.Result{}, err
 		}
 		return ctrl.Result{RequeueAfter: time.Second * 2}, nil
 	}
 
-	if desiredReplicas < currentReplicas-deletedReplicas {
-		// Delete not ready machines first
-		if len(notReadyMachines) > 0 {
-			if err := r.deleteMachine(ctx, &notReadyMachines[0]); err != nil {
-				return ctrl.Result{}, err
-			}
-			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
+	if len(machines) < len(machines)-deletedMachineCount {
+		machineForDeletion := findFirstMachineForDeletion(machines)
+
+		if err := r.deleteMachine(ctx, &machineForDeletion); err != nil {
+			return ctrl.Result{}, err
 		}
+		return ctrl.Result{RequeueAfter: time.Second * 2}, nil
+	}
 
-		// Delete ready machines
-		if len(readyMachines) > 0 {
-			if err := r.deleteMachine(ctx, &readyMachines[0]); err != nil {
-				return ctrl.Result{}, err
+	return ctrl.Result{}, nil
+}
+
+// findFirstMachineForDeletion returns a machine which should be deleted first in case of a scale down.
+// returns unready machines first
+// returns machine which is not keepalived master
+// returns the first machine
+func findFirstMachineForDeletion(machines []yawolv1beta1.LoadBalancerMachine) yawolv1beta1.LoadBalancerMachine {
+	for i := range machines {
+		if !isMachineReady(machines[i]) {
+			return machines[i]
+		}
+	}
+
+	for i := range machines {
+		for _, c := range *machines[i].Status.Conditions {
+			if string(c.Type) == string(helper.KeepalivedMaster) &&
+				string(c.Status) != string(helper.ConditionFalse) {
+				return machines[i]
 			}
-			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
 		}
 	}
 
-	return ctrl.Result{}, nil
+	return machines[0]
```

Comment:
do we need to check length? could this function be called on an empty machine list?